### ISSUE TITLE
Head compaction/snapshot inefficiency

### ISSUE THREAD

**Issue by codesome** [prometheus/tsdb#652]
During snapshot, there are chances of head chunks going beyond the block meta during compaction as we already know and have a fix for. In such case, we add a tombstone to `dranges` here https://github.com/prometheus/tsdb/blob/d230c67aa180850b80ae49e07079f55df1da0502/compact.go#L770

But it is possible for every series to encounter this. Hence the `dranges` will grow linearly with every series iteration and with the same (duplicate) drange (million series = min of million dranges for the last series!). Memory consumption and snapshot time might increase with this.

**Comment by krasi-georgiev** [comment]
Hm, I am not sure I understand the problem here.
Has anyone run into some problems with this? 

We are aware that this will increase memory and overall snapshot time, but can't think of any other way to produce a sane snapshot of the head block.

**Comment by codesome** [comment]
> Has anyone run into some problems with this?

It's in `2.11.0-rc.0` right now, so I don't think anyone would have faced it yet. It was just my speculation.

> but can you think of any other way to produce a sane snapshot of the head block.

The problem that I pointed out is not in the actual solution but in the implementation. We are appending the same drange many times, and the `dranges` might grow very big with the same duplicate `Interval{Mint: meta.MaxTime, Maxt: math.MaxInt64}`, hence the `deleteIterator` will take a lot of time as the Snapshot proceeds (as each series might add the same `Interval` to the slice).

The solution is not adding`Interval{Mint: meta.MaxTime, Maxt: math.MaxInt64}` more than once in `dranges`.

**Comment by krasi-georgiev** [comment]
Aaaa , alright, I understand what you mean now. I am working on something else, now so if you have time I am happy to review a PR.

**Comment by work-chausat** [comment]
i have the same problem, compacting head to disk is really too slow, headIndexReader sorts all series, even a series is not in the time window, the more series in head, the slower to sort series for compaction, and the slower to sort series, the more series in head, finally, the memory is full, oom occurs.

![1111](https://user-images.githubusercontent.com/29337670/61712241-787b1300-ad88-11e9-90d6-0771e9bf1991.png)
![22222](https://user-images.githubusercontent.com/29337670/61712242-7913a980-ad88-11e9-99c4-fd2f40076fae.png)

**Comment by work-chausat** [comment]
1 @ 0x8c21a0 0x4c9bf6 0x4c9e2a 0x4c9ec3 0x4c9e88 0x4c76e9 0x8a5fcf 0x890c28 0x88f9d5 0x88ecfa 0x895f46 0x895889 0x45e281
#	0x8c219f	github.com/prometheus/tsdb.(*headIndexReader).SortedPostings.func1+0x9f	/export/App/gopath/src/github.com/prometheus/tsdb/head.go:1222
#	0x4c9bf5	sort.doPivot_func+0x415										/export/App/go/src/sort/zfuncversion.go:89
#	0x4c9e29	sort.quickSort_func+0x99									/export/App/go/src/sort/zfuncversion.go:143
#	0x4c9ec2	sort.quickSort_func+0x132									/export/App/go/src/sort/zfuncversion.go:148
#	0x4c9e87	sort.quickSort_func+0xf7									/export/App/go/src/sort/zfuncversion.go:145
#	0x4c76e8	sort.Slice+0x128										/export/App/go/src/sort/slice.go:21
#	0x8a5fce	github.com/prometheus/tsdb.(*headIndexReader).SortedPostings+0x39e	/export/App/gopath/src/github.com/prometheus/tsdb/head.go:1222
#	0x890c27	github.com/prometheus/tsdb.(*LeveledCompactor).populateBlock+0x667	/export/App/gopath/src/github.com/prometheus/tsdb/compact.go:713
#	0x88f9d4	github.com/prometheus/tsdb.(*LeveledCompactor).write+0x584		/export/App/gopath/src/github.com/prometheus/tsdb/compact.go:584
#	0x88ecf9	github.com/prometheus/tsdb.(*LeveledCompactor).Write+0x319		/export/App/gopath/src/github.com/prometheus/tsdb/compact.go:490
#	0x895f45	github.com/prometheus/tsdb.(*DB).compact+0x205				/export/App/gopath/src/github.com/prometheus/tsdb/db.go:440
#	0x895888	github.com/prometheus/tsdb.(*DB).run+0x268				/export/App/gopath/src/github.com/prometheus/tsdb/db.go:362


it is too slow, gc for head is postponed, and the series number increases rapidlly, the next compaction will be  slower than this time. Vicious circle!!!!!

**Comment by work-chausat** [comment]
i added "level.Info(h.head.logger).Log("msg", "sort series", "length", len(series))" in headIndexReader.SortedPostings.

`func (h *headIndexReader) SortedPostings(p index.Postings) index.Postings {
	series := make([]*memSeries, 0, 128)

	// Fetch all the series only once.
	for p.Next() {
		s := h.head.series.getByID(p.At())
		if s == nil {
			level.Debug(h.head.logger).Log("msg", "looked up series not found")
		} else {
			series = append(series, s)
		}
	}
	if err := p.Err(); err != nil {
		return index.ErrPostings(errors.Wrap(err, "expand postings"))
	}
	level.Info(h.head.logger).Log("msg", "sort series", "length", len(series))
	sort.Slice(series, func(i, j int) bool {
		return labels.Compare(series[i].lset, series[j].lset) < 0
	})

	// Convert back to list.
	ep := make([]uint64, 0, len(series))
	for _, p := range series {
		ep = append(ep, p.ref)
	}
	return index.NewListPostings(ep)
}`

then the log is like this:

`level=info time=2019-07-23T21:38:24.596610954 caller=head.go:1221 msg="sort series" length=1101589
level=info time=2019-07-23T21:38:29.662662861 caller=compact.go:499 msg="write block" mint=1563888950260 maxt=1563888960000 ulid=01DGFGVXGJ1XKXTW32C66H6CKA duration=5.260009575s
level=info time=2019-07-23T21:38:31.588181845 caller=head.go:545 msg="head GC completed" duration=1.553269161s
level=info time=2019-07-23T21:38:44.151516687 caller=head.go:1221 msg="sort series" length=2961010
level=info time=2019-07-23T21:39:15.664188588 caller=compact.go:499 msg="write block" mint=1563888960000 maxt=1563889020000 ulid=01DGFGWG7KSRWRRVPBGZY8TWPX duration=32.093063613s
level=info time=2019-07-23T21:39:25.587715375 caller=head.go:545 msg="head GC completed" duration=8.340222755s
level=info time=2019-07-23T21:39:26.882991554 caller=head.go:1221 msg="sort series" length=332933
level=info time=2019-07-23T21:39:29.553622655 caller=compact.go:499 msg="write block" mint=1563889020000 maxt=1563889080000 ulid=01DGFGXTF2Q2W5TGR1YQYWF2KF duration=2.735273022s
level=info time=2019-07-23T21:39:30.490586183 caller=head.go:545 msg="head GC completed" duration=511.582331ms
level=info time=2019-07-23T21:39:43.568209254 caller=head.go:1221 msg="sort series" length=333288
level=info time=2019-07-23T21:39:46.360807189 caller=compact.go:499 msg="write block" mint=1563889080000 maxt=1563889140000 ulid=01DGFGYARJX61MNXE6NG34BMTS duration=2.85384871s
level=info time=2019-07-23T21:39:47.313954542 caller=head.go:545 msg="head GC completed" duration=616.957528ms
level=info time=2019-07-23T21:40:22.616105275 caller=head.go:1221 msg="sort series" length=651706
level=info time=2019-07-23T21:40:28.676260782 caller=compact.go:499 msg="write block" mint=1563889140000 maxt=1563889200000 ulid=01DGFGZGT3Q6315FC6T7Q042G8 duration=6.208455159s
level=info time=2019-07-23T21:40:30.593320941 caller=head.go:545 msg="head GC completed" duration=1.354086362s
level=info time=2019-07-23T21:41:23.597320358 caller=head.go:1221 msg="sort series" length=1044285
level=info time=2019-07-23T21:41:34.115490696 caller=compact.go:499 msg="write block" mint=1563889200000 maxt=1563889260000 ulid=01DGFH1CABP5E7ZEF24K2EJFR4 duration=10.711680145s
level=info time=2019-07-23T21:41:36.867320626 caller=head.go:545 msg="head GC completed" duration=2.094125246s
level=info time=2019-07-23T21:43:11.233927147 caller=head.go:1221 msg="sort series" length=4879288
level=info time=2019-07-23T21:44:37.979903022 caller=compact.go:499 msg="write block" mint=1563889260000 maxt=1563889320000 ulid=01DGFH4MQWYJB7MAYX3DNB4HEJ duration=1m27.647885153s
level=info time=2019-07-23T21:45:27.478879693 caller=head.go:545 msg="head GC completed" duration=41.306104387s
level=info time=2019-07-23T21:45:43.571891323 caller=head.go:1221 msg="sort series" length=16982472`

**Comment by codesome** [comment]
@work-chausat More info is needed an analyze the issue. I could only deduce that you had 1.1M head series. Could you open another GitHub issue with more details (like the Y axis of your graph is not visible, memory used by prometheus and the system capacity, prometheus version, time taken for compaction).
